pyogrio/_io.pyx

```diff
@@ -1722,6 +1722,11 @@ def ogr_open_arrow(
         if columns is not None:
             # Fields are matched exactly by name, duplicates are dropped.
             ignored_fields = list(set(fields[:, 2]) - set(columns))
+
+            # Find index of each field in columns, and only keep those
+            idx = np.intersect1d(fields[:, 2], columns, return_indices=True)[1]
```

Comment:
I think the intersect1d might give a sorted result. While this will not be problematic I assume in practice (we loop over it, but still using the name and not position in the list), let's maybe preserve the order in the file.

Reply:
```suggestion
            idx = np.sort(np.intersect1d(fields[:, 2], columns, return_indices=True)[1])
```

might be sufficient?